lib/Doctrine/DBAL/Driver/Connection.php

```diff
@@ -15,62 +14,80 @@
 {
     /**
      * Prepares a statement for execution and returns a Statement object.
+     *
+     * @param string $sql The SQL query to prepare.
+     *
+     * @return Statement The prepared statement.
      */
     public function prepare(string $sql) : Statement;
 
     /**
      * Executes an SQL statement, returning a result set as a Statement object.
      *
+     * @param string $sql The SQL query to execute.
+     *
+     * @return ResultStatement The result statement.
+     *
      * @throws DBALException
      */
     public function query(string $sql) : ResultStatement;
 
     /**
      * Quotes a string for use in a query.
      *
-     * @param mixed $input
-     * @param int   $type
+     * @param string $input The parameter to quote.
      *
-     * @return mixed
+     * @return string The quoted string.
      */
-    public function quote($input, $type = ParameterType::STRING);
+    public function quote(string $input) : string;
 
     /**
      * Executes an SQL statement and return the number of affected rows.
      *
+     * @param string $statement The SQL query to execute.
+     *
+     * @return int The number of affected rows.
+     *
      * @throws DBALException
      */
     public function exec(string $statement) : int;
 
     /**
      * Returns the ID of the last inserted row or sequence value.
      *
-     * @param string|null $name
+     * If a sequence name was not specified, lastInsertId() returns a string representing the row ID of the last row
+     * that was inserted into the database.
+     *
+     * If a sequence name was specified, lastInsertId() returns a string representing the last value retrieved from the
+     * specified sequence object.
      *
-     * @return string
+     * This method throws a DriverException if a value cannot be returned, in particular when:
+     *
+     * - no sequence name was provided, but the platform requires one, or the last statement did not return an identity;
```

Comment:
> but the platform requires one

It's a driver-specific API, so "driver" seems more suitable here than "platform".

Reply:
> the last statement did not return an identity

Do we really want an exception here? In https://github.com/doctrine/dbal/pull/3335#discussion_r234417100, I answered "exception" to all cases without having noticed that the question contained multiple scenarios. The current implementation of this case doesn't seem to be consistent across drivers (e.g. compare SQL Anywhere and IBM DB2). Which means it's not tested and not documented.

I feel like the last one is not an exceptional situation, but a candidate for `null`. What do you think?

Reply:
> Do you have a use case where this would actually be useful?

No.